### ISSUE TITLE
Support for multiple MCP sources for endpoints

### ISSUE THREAD

**Issue by Jpatel86** [istio/istio#15417]
(NOTE: This is used to report product bugs:
  To report a security vulnerability, please visit <https://istio.io/about/security-vulnerabilities/>
  To ask questions about how to use Istio, please visit <https://discuss.istio.io>
)

**Bug description**
For endpoints/service discovery, multiple MCP sources should be supported in multi cluster environment.

To sync service discovery from external sources, example consul(https://github.com/costinm/istio-consul), MCP sources for endpoints should be available and will sync the updates with Pilot/envoy.
**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Service discovery integration between istio and consul using MCP endpoints.  
**Steps to reproduce the bug**

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
Istio 1.x.x
**How was Istio installed?**
helm templates
**Environment where bug was observed (cloud vendor, OS, etc)**
Ubuntu
Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by costinm** [comment]
@nmittler 

I believe multiple MCP sources are supported in the config - the problem seems to be that we want the first ( or one ) to be used for config objects, and the others to be used for ServiceEntry only.

Or to adjust the config to have a set of 'endpoint' sources, separate from the config sources.

One option worth considering here - given that there are other implementations of consul adapters exposing endpoints as EDS - is to have one adapter in pilot that can retrieve endpoints from an EDS source. Long term we are considering converging MCP and XDS ( by extending XDS ), so effectively this will allow Pilot to aggregate endpoints expressed as either EDS or ServiceEntry.

**Comment by nmittler** [comment]
@costinm @ozevren @Nino-K 

Is the problem simply that Galley publishes `ServiceEntry` resources in the same collection as everything else? So if you subscribe to multiple galleys, you get duplicates of all of the resources?

Perhaps a first step would be to put `ServiceEntry` into a separate collection that Pilot can consume selectively?

**Comment by costinm** [comment]
I am not sure what the problem is - I think setting up ( something about the sortedCollections flags). 
This is using Consul - which does its own readiness, so it doesn't have the issues we have with k8s, and 
having the ServiceEntry sent in the same collections should be fine. 

So documentation or some existing test that configures Pilot with 2 MCPs would probably unblock this.
And making sure that we only ask for ServiceEntries - not all other collections - for non-primary MCP servers.

**Comment by nmittler** [comment]
@ayj @ozevren @Nino-K  it appears that Pilot always consumes all types from all MCP servers that it's configured with. This doesn't seem right to me. I would think we should be able to select the collections we receive from each server in [the mesh config](https://github.com/istio/api/blob/master/mesh/v1alpha1/config.proto#L319).  That would effectively just require adding a list of "collections" to this proto (empty means everything).

WDYT?

**Comment by Nino-K** [comment]
The PR that is in flight right now for synthetic service entries does exactly that, it leaves the main controller for all the the config types except ServiceEntries (synthetic)/ endpoints and one dedicated controller just for ServiceEntries/endpoints. Also the dedicated controller can be horizontally scaled by providing multiple source addresses. As for configuring pilot to only support specific types we had this conversation in the past [here](https://github.com/istio/istio/issues/13559), there are a few ways that it can be tackled, but nevertheless I agree, I think pilot should be able to only listen and support configured types.

**Comment by nmittler** [comment]
@Nino-K 

> The PR that is in flight right now for synthetic service entries does exactly that, it leaves the main controller for all the the config types except ServiceEntries (synthetic)/ endpoints and one dedicated controller just for ServiceEntries/endpoints

That's not exactly what we want here ... this is a bit more generic.  In this case we would want *real* ServiceEntry resources to be served from all MCP servers (for endpoints), but the rest of the configuration to only come from a single (i.e. *this* cluster's) MCP server.

Making this change would (I think) dovetail nicely with your PR ... in fact it should make some of your plumbing in Pilot a bit more generic.

**Comment by Nino-K** [comment]
@nmittler The idea of having a dedicated controller(s) just for endpoints and one controller for other config is mainly what I meant. 

>Making this change would (I think) dovetail nicely with your PR ... in fact it should make some of your plumbing in Pilot a bit more generic.

+1

**Comment by nmittler** [comment]
@Jpatel86 re-reading the description, it isn't clear to me what the issue is. IIUC the mesh config already supports multiple galleys, and each galley can serve ServiceEntry. Are you saying that it's not working?

**Comment by Jpatel86** [comment]
@nmittler The issue I am working on, service discovery sync between consul and Istio. I came across the consul-istio adapter which @costinm created(https://github.com/costinm/istio-consul), it listens on local consul agent and expose MCP on specific port(15098).
If I talk about this specific software, pilot detects this MCP server but it's  unable to synchronize endpoints. I think it's missing command line flags ,sortedCollections , implementation.

**Comment by nmittler** [comment]
@Jpatel86 thanks for clarifying

@costinm do we have any integration tests for the consul integration? I took a quick look through and didn't see anything.
